res/controllers/Traktor-Kontrol-S2-MK2-hid-scripts.js

```diff
@@ -0,0 +1,1389 @@
+  
+  MessageShort.setCallback("[EffectRack1_EffectUnit1]", "!effect_focus_button", this.effectFocusButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit1", "!effectbutton1", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit1", "!effectbutton2", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit1", "!effectbutton3", this.effectButton);
+ 
+  MessageShort.setCallback("[EffectRack1_EffectUnit2]", "!effect_focus_button", this.effectFocusButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit2", "!effectbutton1", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit2", "!effectbutton2", this.effectButton);
+  MessageShort.setCallback("[EffectRack1_EffectUnit2", "!effectbutton3", this.effectButton);
+  
+  MessageShort.setCallback("[Master]", "maximize_library", this.toggleButton);
+  
+  MessageShort.setCallback("[Microphone]", "talkover", this.toggleButton);
+  
+  engine.makeConnection("[EffectRack1_EffectUnit1]", "show_parameters", TraktorS2MK2.onShowParametersChange);
+  engine.makeConnection("[EffectRack1_EffectUnit2]", "show_parameters", TraktorS2MK2.onShowParametersChange);
+  
+  this.controller.registerInputPacket(MessageShort);
+
+  // Most items in the long message are controls that go from 0-4096.
+  // There are also some 4 bit encoders.
+  MessageLong.addControl("[Channel1]", "rate", 0x07, "H");
+  MessageLong.addControl("[Channel2]", "rate", 0x09, "H");
+  engine.softTakeover("[Channel1]", "rate", true);
+  engine.softTakeover("[Channel2]", "rate", true);
+  MessageLong.addControl("[Channel1]", "!left_encoder", 0x01, "B", 0x0F, undefined, true);
+  MessageLong.addControl("[Channel2]", "!left_encoder", 0x02, "B", 0xF0, undefined, true);
+  MessageLong.setCallback("[Channel1]", "!left_encoder", this.leftEncoder);
+  MessageLong.setCallback("[Channel2]", "!left_encoder", this.leftEncoder);
+  MessageLong.addControl("[Channel1]", "!right_encoder", 0x01, "B", 0xF0, undefined, true);
+  MessageLong.addControl("[Channel2]", "!right_encoder", 0x03, "B", 0x0F, undefined, true);
+  MessageLong.setCallback("[Channel1]", "!right_encoder", this.rightEncoder);
+  MessageLong.setCallback("[Channel2]", "!right_encoder", this.rightEncoder);
+
+  MessageLong.addControl("[EffectRack1_EffectUnit1]", "mix", 0x17, "H");
+  MessageLong.addControl("[EffectRack1_EffectUnit1", "!effectknob1", 0x19, "H");
+  MessageLong.setCallback("[EffectRack1_EffectUnit1", "!effectknob1", this.effectKnob);
+  MessageLong.addControl("[EffectRack1_EffectUnit1", "!effectknob2", 0x1B, "H");
+  MessageLong.setCallback("[EffectRack1_EffectUnit1", "!effectknob2", this.effectKnob);
+  MessageLong.addControl("[EffectRack1_EffectUnit1", "!effectknob3", 0x1D, "H");
+  MessageLong.setCallback("[EffectRack1_EffectUnit1", "!effectknob3", this.effectKnob);
+
+  MessageLong.addControl("[EffectRack1_EffectUnit2]", "mix", 0x1F, "H");
+  MessageLong.addControl("[EffectRack1_EffectUnit2", "!effectknob1", 0x21, "H");
```

Comment:
same here, forgot the closing bracket?

Reply:
That was intentional to make the string manipulation a bit easier. I'll add a comment.